### ISSUE TITLE
Are there plans to move workflows into their own repos?

### ISSUE THREAD

**Issue by rpetit3** [StaPH-B/staphb_toolkit#60]
Hi everyone, 

The toolkit is great! I'm just curious if there were plans to move the workflows to their own repo. 

My thinking is the staphb_toolkit could import them, or in the case of Nextflow-based ones just run from the repo directly. I also think it might be easier to submit contributions.

Thanks!
Robert

**Comment by k-florek** [comment]
@rpetit3 I agree completely!! This is the intent for the next version of the toolkit but it will take a bit of time to get there with everything else going on!

**Comment by erinyoung** [comment]
I agree! I was realizing that as we add newborn screening and other workflows to the toolkit, this repo may get too big and unwieldy.

**Comment by rsettlage** [comment]
Another tweak you may choose to make, if you combine the containers within a workflow, you will realize space savings.  For an analysis of lots of samples via cloud, this could result in some significant cost savings.  When I did this with the Monroe PE containers, the total space went from >2.5 GB to <1.1 GB.  I am not sure what the ramifications would be when using NextFlow.  For my HPC-like cloud workflow, this was a good thing as I pay for space.

Bob

**Comment by k-florek** [comment]
@erinyoung Even with the 7 workflows the toolkit has it is already unwieldy! I have a few ideas on how I would like to implement this but I might ask you for some input down the road.

@rsettlage Interesting thought, and I could see that being helpful for running the workflows locally. Generally speaking it is a best practice to keep docker images small and light and there are a lot of factors that go into image/container optimization. For example a lot of the StaPH-B containers utilize a similar base image and since images exist on a disk as layers this reduces the number of different layers needed. Additionally, Cloud workflows and HPC systems often run jobs in an ephemeral approach where jobs are split across multiple instances that are often removed or discarded after use so the image is in essence downloaded for each instance. Thus keeping the images small and light and specific to their individual tasks speeds up execution time since the process won't need to download a large image.

**Comment by rsettlage** [comment]
I don't have much experience using NextFlow, but for HPC, to avoid a lot of transfers, job wait time etc, I pull the single combined container to the compute node (really a bare metal cloud VM) along with data for a single sample, go through the entire pipeline and then transfer results back.  You could imagine a workflow where each step was a separate job to optimize the core count at each step, perhaps this is the NextFlow way?, but these jobs are so tiny that is more wasteful in the start/stop/transfer on HPC, such that IME it is better to do it in a single go.  So these jobs are specific to the sample, not the task.

Again, pulling each of the 7(?) containers individually resulted in >2.5 GB of total container space while the combined container was ca. 1 GB.  I assumed this was due to 1 copy of the OS vs 7.

I would love to compare compute resources and costs doing it other ways...if only we had infinite time ...

**Comment by erinyoung** [comment]
> You could imagine a workflow where each step was a separate job to optimize the core count at each step, perhaps this is the NextFlow way?

@rsettlage Yes, this is the nextflow way, or any workflow manager way. I really recommend you look into a workflow manager, such as nextflow or snakemake (or even wdl - I won't judge) to manage your workflows. Most work well with HPC environments. It'll make your life easier. I promise.

**Comment by rsettlage** [comment]
It sure would have when asked to duplicate the Monroe pipeline.  ;)  I have used Snakemake.

But, ultimately, it is about the environment available and the resources consumed.  So, time to compare workflows.  I will see what it would take to get NF going in the cloud env I have available.

Do you have any performance/resource utilization numbers available?  I suppose one would need to either start with the same samples or standardize on an instrument/sample count/read output set to get something somewhat comparable.

**Comment by fanninpm** [comment]
Nextflow has a way to run a workflow straight from its GitHub repository. This can be adapted to Cecret by adding the `-main-script` flag to [this invocation](https://github.com/StaPH-B/staphb_toolkit/blob/2a2c2013eb193228f5086dfd74e7cc5877251e90/staphb_toolkit/toolkit_workflows.py#L693) and pointing it to the correct script. See [here](https://www.nextflow.io/docs/latest/sharing.html) for how that can work.

**Comment by k-florek** [comment]
@fanninpm Thanks for pointing this out! That is likely the route I will go.

**Comment by k-florek** [comment]
Workflows are now sourced from their own repositories in version 2 of the toolkit!
